src/dataframe/dataframe.jl

```diff
@@ -697,15 +697,28 @@ end
 ##
 ##############################################################################
 
-# hcat! for 2 arguments
+# hcat! for 2 arguments, only a vector or a data frame is allowed
 function hcat!(df1::DataFrame, df2::AbstractDataFrame)
     u = add_names(index(df1), index(df2))
     for i in 1:length(u)
         df1[u[i]] = df2[i]
     end
     return df1
 end
+
+# definition required to avoid hcat! ambiguity
+function hcat!(df1::DataFrame, df2::DataFrame)
+    invoke(hcat!, Tuple{DataFrame, AbstractDataFrame}, df1, df2)
+end
+
 hcat!(df::DataFrame, x::AbstractVector) = hcat!(df, DataFrame(Any[x]))
+hcat!(x::AbstractVector, df::DataFrame) = hcat!(DataFrame(Any[x]), df)
+function hcat!(x, df::DataFrame)
```

Comment:
"Abstract". Same below.

Reply:
I would leave it as is. I thought it was a design decision to have `hcat!` handle `DataFrame` and that is why it is in *dataframe.jl*.
`hcat!` is not exported and those two cases actually will be called by those functions in *abstractdataframe.jl*:
```
Base.hcat(df::AbstractDataFrame, x) = hcat!(df[:, :], x)
Base.hcat(x, df::AbstractDataFrame) = hcat!(x, df[:, :])
```
which make sure to create a `DataFrame`. Any comment on it (I do not know why it was designed like this and I do not want to break the original intention)?

Reply:
No, I meant you should fix "AbstaractVector" on the next line, not sure why it ended up on this line.

Reply:
ah - those typos. fixed